File: 3437-maximum-total-damage-with-spell-casting/maximum-total-damage-with-spell-casting.py
class Solution:
    def helper(self, idx, arr, prev, n, dp):
        if idx < 0: return 0
        if dp[idx][prev] != -1: return dp[idx][prev]
        not_take = self.helper(idx-1,arr, prev, n, dp)
        take = -1
        if prev == n or arr[prev][0] - arr[idx][0] > 2:
            take = arr[idx][1] + self.helper(idx-1,arr,idx,n, dp)
        dp[idx][prev] = max(take,not_take)
        return dp[idx][prev]
        
    def maximumTotalDamage(self, power: List[int]) -> int:
        c = Counter(power)
        arr =  sorted(c.keys())
        if not arr: return 0
        total = [i*c[i] for i in arr]
        n = len(arr)
        dp = [0]*n
        dp[0] = total[0]
        for i in range(1,n):
            not_take = max(total[i], dp[i-1])
            take = 0
            j = i-1
            while j >= 0 and arr[i]-arr[j] <= 2:
                j-= 1
            if j != -1:
                take = total[i] + dp[j]
            dp[i] = max(not_take, take)

        return dp[n-1] 

File: 3437-maximum-total-damage-with-spell-casting/maximum-total-damage-with-spell-casting.cpp
class Solution {
public:
    long long maximumTotalDamage(vector<int>& power) {
        unordered_map<int,int> Freq;
        for (int d: power) Freq[d]++;
        int n = 0;
        vector<int> vals;
        for (auto [k,v]: Freq){
            n++;
            vals.push_back(k);
        }
        if (n==0) return 0;
        sort(vals.begin(), vals.end());
        vector<long long> totals;
        for (int i=0; i<n; i++){
            totals.push_back(1LL*vals[i]*Freq[vals[i]]);
        }
        vector<long long> dp(n,0);
        dp[0] = totals[0];
        long long prev, no_prev;
        int j;
        for (int idx=1; idx<n; idx++){
            no_prev = std::max(dp[idx-1],totals[idx]);
            prev = 0;
            j = idx-1;
            while (j>=0 && vals[idx]-vals[j]<=2) j--;
            if (j != -1){
                prev = dp[j] + totals[idx];
            }
            dp[idx] = std::max(prev, no_prev);
        }
        return dp[n-1];
    }
};